README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# har18: Implementation of Harrison & Pearce’s (2018) Harmonicity Model

[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![Travis build
status](https://travis-ci.org/pmcharrison/har18.svg?branch=master)](https://travis-ci.org/pmcharrison/har18)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/pmcharrison/har18?branch=master&svg=true)](https://ci.appveyor.com/project/pmcharrison/har18)
[![Coverage
status](https://coveralls.io/repos/github/pmcharrison/har18/badge.svg)](https://coveralls.io/r/pmcharrison/har18?branch=master)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2545752.svg)](https://doi.org/10.5281/zenodo.2545752)

Harmonicity is an important predictor of a chord’s consonance. This
package provides methods for estimating a chord’s harmonicity, after
Harrison & Pearce (2018) and Milne (2013).

## Installation

You can install this package from GitHub:

``` r
if (!require(devtools)) install.packages("devtools")
devtools::install_github("pmcharrison/har18")
```

## Example usage

A major chord is more harmonic than a diminished chord, as expected:

``` r
library(har18)

pc_harmonicity("0 4 7")
#> [1] 0.9373149
pc_harmonicity("0 3 6")
#> [1] 0.7225033
```

By default, the package implements the method of Harrison & Pearce
(2018). Setting the option `method = "peak"` reproduces the method of
Milne (2013).

``` r
pc_harmonicity("0 4 7", method = "peak")
#> [1] 0.7170107
pc_harmonicity("0 3 6", method = "peak")
#> [1] 0.5896047
```

## References

Harrison, P. M. C., & Pearce, M. T. (2018). An energy-based generative
sequence model for testing sensory theories of Western harmony. In
Proceedings of the 19th International Society for Music Information
Retrieval Conference. Paris, France. Retrieved from
<https://arxiv.org/abs/1807.00790>

Milne, A. J. (2013). A computational model of the cognition of tonality.
The Open University, Milton Keynes, England.


NEWS.md
# har18 0.2.1

* Rewrite sweep_template in C++ for speed.

# har18 0.2.0

* Add a `NEWS.md` file to track changes to the package.
* Generalise `kl_div_from_uniform` to work with arbitrary objects of class `smooth_spectrum`.


tests/testthat.R
library(testthat)
library(har18)

test_check("har18")


tests/testthat/test-cpp.R
context("test-cpp")

test_that("get_index", {
  x <- c(0, 1, 2, 3)

  expect_equal(get_index(x, index = 0, offset = 0), 0)
  expect_equal(get_index(x, index = 1, offset = 0), 1)
  expect_equal(get_index(x, index = 2, offset = 0), 2)
  expect_equal(get_index(x, index = 3, offset = 0), 3)

  expect_equal(get_index(x, index = 0, offset = 1), 1)
  expect_equal(get_index(x, index = 1, offset = 1), 2)
  expect_equal(get_index(x, index = 2, offset = 1), 3)
  expect_equal(get_index(x, index = 3, offset = 1), 0)

  expect_equal(get_index(x, index = 0, offset = 2), 2)
  expect_equal(get_index(x, index = 1, offset = 2), 3)
  expect_equal(get_index(x, index = 2, offset = 2), 0)
  expect_equal(get_index(x, index = 3, offset = 2), 1)
})

test_that("cosine_similarity", {
  x <- c(12, 25, -40, 20, 49)
  y <- c(-40, 30, 10, 5, -30)

  expect_equal(
    cosine_similarity_cpp(x, y, offset = 0), # <-- C++ version
    cosine_similarity(x, y), # <-- R version
  )

  expect_equal(
    cosine_similarity_cpp(x, y, offset = 1),
    cosine_similarity(x[c(2:5, 1)], y),
  )

  expect_equal(
    cosine_similarity_cpp(x, y, offset = 2),
    cosine_similarity(x[c(3:5, 1:2)], y),
  )

  expect_equal(
    cosine_similarity_cpp(x, y, offset = 3),
    cosine_similarity(x[c(4:5, 1:3)], y),
  )
})

test_that("sweep_template", {
  x <- c(-5, 20, 15, 35, 40, 20)
  y <- c(20, 15, 27, 40, 10, 15)

  expect_equal(
    sweep_template(x, y, legacy = TRUE),
    sweep_template_cpp(x, y)
  )
})


tests/testthat/test-peak.R
context("test-peak")

library(hrep)
library(magrittr)

test_that("example", {
  c(0, 0, 3, 4, 1) %>%
    .milne_pc_spectrum() %>%
    peak %>%
    expect_equal(4)
})

test_that("invariance to doubling the number of bins", {
  for (i in 1:10) {
    spec1 <- rnorm(10) %>% .milne_pc_spectrum()
    spec2 <- rep(as.numeric(spec1), each = 2) %>% .milne_pc_spectrum()
    expect_equal(
      peak(spec1),
      peak(spec2)
    )
  }
})

test_that("more pointy distributions get higher peaks", {
  x <- seq(from = -100, to = 100)
  pointy <- dnorm(x, sd = 1)
  less_pointy <- dnorm(x, sd = 4)
  expect_gt(
    peak(.milne_pc_spectrum(pointy)),
    peak(.milne_pc_spectrum(less_pointy))
  )
})


R/features.R
#' Peak
#'
#' Gets the peak value of an object.
#' @param x Object to analyse.
#' @return The object's peak value, as a numeric scalar.
#' @rdname peak
#' @export
peak <- function(x) {
  UseMethod("peak")
}

#' @rdname peak
#' @export
peak.milne_pc_spectrum <- function(x) {
  max(as.numeric(x))
}

#' Kullback-Leibler divergence from uniform
#'
#' Gets the Kullback-Leibler divergence of a provided distribution
#' from a uniform distribution.
#' @param x Input distribution.
#' @return The Kullback-Leibler divergence from a uniform distribution
#' to the input distribution.
#' @rdname kl_div_from_uniform
#' @export
kl_div_from_uniform <- function(x) {
  UseMethod("kl_div_from_uniform")
}

#' @rdname kl_div_from_uniform
#' @export
kl_div_from_uniform.smooth_spectrum <- function(x) {
  # Construct a probability vector, where each bin corresponds to
  # the probability of a discrete event
  x <- as.numeric(x)
  probs <- x / sum(x)
  n <- length(probs)
  uniform_probs <- 1 / n
  non_zero_probs <- probs[probs > 0]
  sum(
    non_zero_probs * log(non_zero_probs / uniform_probs, base = 2)
  )
}


R/harmonicity.R
#' @export
harmonicity <- function(x,
                        method = "kl",
                        type = "pc",
                        num_harmonics = 12,
                        rho = 0.75,
                        sigma = 6.83,
                        array_dim = 1200) {
  UseMethod("harmonicity")
}

#' @export
harmonicity.default <- function(x, type = "pc", ...) {
  x <- if (type == "pc") {
    pc_set(x)
  } else if (type == "pi") {
    stop("pitch-based harmonicity is not currently supported")
    pi_chord(x)
  } else {
    stop("unrecognised type")
  }
  harmonicity(x, type = type, ...)
}

#' @export
harmonicity.pc_set <- function(x,
                               method = "kl",
                               type = "pc",
                               num_harmonics = 12,
                               rho = 0.75,
                               sigma = 6.83,
                               array_dim = 1200,
                               ...) {
  checkmate::qassert(method, "S1")
  checkmate::qassert(type, "S1")

  x <- hrep::pc_smooth_spectrum(x,
                                num_harmonics = num_harmonics,
                                rho = rho,
                                sigma = sigma,
                                array_dim = array_dim)
  harmonicity(x,
              method = method,
              type = type,
              num_harmonics = num_harmonics,
              rho = rho,
              sigma = sigma)
}

#' @export
harmonicity.pc_smooth_spectrum <- function(x,
                                           method = "kl",
                                           type = "pc",
                                           num_harmonics = 12,
                                           rho = 0.75,
                                           sigma = 6.83) {
  checkmate::qassert(method, "S1")
  checkmate::qassert(type, "S1")

  # One day we could generalise the method from pitch classes to pitches.
  # This is not implemented yet, though.
  stopifnot(type == "pc")

  virtual <- sweep_harmonic_template(x,
                                     num_harmonics = num_harmonics,
                                     rho = rho,
                                     sigma = sigma,
                                     array_dim = array_dim)
  if (method == "kl") {
    kl_div_from_uniform(virtual)
  } else if (method == "peak") {
    peak(virtual)
  } else stop("unrecognised method")
}


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

// mod
double mod(double x, int base);
RcppExport SEXP _har18_mod(SEXP xSEXP, SEXP baseSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< double >::type x(xSEXP);
    Rcpp::traits::input_parameter< int >::type base(baseSEXP);
    rcpp_result_gen = Rcpp::wrap(mod(x, base));
    return rcpp_result_gen;
END_RCPP
}
// get_index
double get_index(NumericVector& x, int index, int offset);
RcppExport SEXP _har18_get_index(SEXP xSEXP, SEXP indexSEXP, SEXP offsetSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericVector& >::type x(xSEXP);
    Rcpp::traits::input_parameter< int >::type index(indexSEXP);
    Rcpp::traits::input_parameter< int >::type offset(offsetSEXP);
    rcpp_result_gen = Rcpp::wrap(get_index(x, index, offset));
    return rcpp_result_gen;
END_RCPP
}
// cosine_similarity_cpp
double cosine_similarity_cpp(NumericVector& x, NumericVector& template_, int offset);
RcppExport SEXP _har18_cosine_similarity_cpp(SEXP xSEXP, SEXP template_SEXP, SEXP offsetSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericVector& >::type x(xSEXP);
    Rcpp::traits::input_parameter< NumericVector& >::type template_(template_SEXP);
    Rcpp::traits::input_parameter< int >::type offset(offsetSEXP);
    rcpp_result_gen = Rcpp::wrap(cosine_similarity_cpp(x, template_, offset));
    return rcpp_result_gen;
END_RCPP
}
// sweep_template_cpp
NumericVector sweep_template_cpp(NumericVector& x, NumericVector& template_);
RcppExport SEXP _har18_sweep_template_cpp(SEXP xSEXP, SEXP template_SEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericVector& >::type x(xSEXP);
    Rcpp::traits::input_parameter< NumericVector& >::type template_(template_SEXP);
    rcpp_result_gen = Rcpp::wrap(sweep_template_cpp(x, template_));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_har18_mod", (DL_FUNC) &_har18_mod, 2},
    {"_har18_get_index", (DL_FUNC) &_har18_get_index, 3},
    {"_har18_cosine_similarity_cpp", (DL_FUNC) &_har18_cosine_similarity_cpp, 3},
    {"_har18_sweep_template_cpp", (DL_FUNC) &_har18_sweep_template_cpp, 2},
    {NULL, NULL, 0}
};

RcppExport void R_init_har18(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/sweep-harmonic-template.R
#' Sweep harmonic template
#'
#' Sweeps a harmonic template over an input spectrum.
#' @param x Object to analyse.
#' @param num_harmonics See \code{\link{pc_harmonicity}}.
#' @param rho See \code{\link{pc_harmonicity}}.
#' @param sigma See \code{\link{pc_harmonicity}}.
#' @param array_dim See \code{\link{pc_harmonicity}}.
#' @param ... Arguments passed to specific methods.
#' @return An object of class \code{\link[hrep]{milne_pc_spectrum}},
#' identifying each pitch class with a perceptual weight
#' corresponding to its harmonic template fit.
#' @rdname sweep_harmonic_template
#' @export
sweep_harmonic_template <- function(x,
                                    num_harmonics = 12,
                                    rho = 0.75,
                                    sigma = 6.83,
                                    array_dim = 1200,
                                    ...) {
  UseMethod("sweep_harmonic_template")
}

#' @rdname sweep_harmonic_template
#' @export
sweep_harmonic_template.pc_set <- function(x,
                                           num_harmonics = 12,
                                           rho = 0.75,
                                           sigma = 6.83,
                                           array_dim = 1200,
                                           ...) {
  hrep::milne_pc_spectrum(x,
                          num_harmonics = num_harmonics,
                          rho = rho,
                          sigma = sigma,
                          array_dim = array_dim) %>%
    sweep_harmonic_template(num_harmonics = num_harmonics,
                            rho = rho,
                            sigma = sigma)
}

#' @rdname sweep_harmonic_template
#' @export
sweep_harmonic_template.milne_pc_spectrum <- function(x,
                                                      num_harmonics = 12,
                                                      rho = 0.75,
                                                      sigma = 6.83,
                                                      ...) {
  x <- as.numeric(x)
  array_dim <- length(x)
  template <- hrep::milne_pc_spectrum(hrep::pc_set(0),
                                      array_dim = array_dim,
                                      num_harmonics = num_harmonics,
                                      rho = rho,
                                      sigma = sigma)
  res <- sweep_template(x, template)

  hrep::.milne_pc_spectrum(res)
}

#' Sweep template
#'
#' Sweeps a circular template over a circular vector
#' and computes the cosine similarity at each possible offset.
#'
#' @param x
#' (Numeric vector)
#' The vector to be swept over.
#'
#' @param template
#' (Numeric vector)
#' The template to sweep over \code{x}.
#' Should have the same dimensionality as \code{x}.
#'
#' @param legacy
#' (Logical scalar)
#' Whether to use the legacy R implementation
#' (default = \code{FALSE}).
#' Otherwise the faster C++ implementation is used.
#'
#' @export
sweep_template <- function(x, template, legacy = FALSE) {
  if (!legacy) {
    return(sweep_template_cpp(x, template))
  }

  array_dim <- length(x)
  res <- numeric(array_dim)

  for (i in seq_len(array_dim)) {
    indices <- 1 + (seq(from = i - 1, length.out = array_dim) %% array_dim)
    res[i] <- cosine_similarity(template, x[indices])
  }

  res
}


R/pc-harmonicity.R
#' Pitch-class harmonicity
#'
#' Gets the pitch-class harmonicity of an input sonority, after
#' \insertCite{Harrison2018;textual}{har18} and
#' \insertCite{Milne2013;textual}{har18}.
#' @param x Object to analyse.
#' @param method (Character scalar) Method to use.
#' * \code{"kl"} (default) delivers the Kullback-Leibler method of
#' \insertCite{Harrison2018;textual}{har18}.
#' * \code{"peak"} delivers the peak-value method of
#' \insertCite{Milne2013;textual}{har18}.
#' @param num_harmonics (Integerish scalar)
#' Number of harmonics to use when expanding tones into their implied harmonics,
#' and when defining the harmonic template
#' (including the fundamental frequency).
#' Defaults to 12, after
#' \insertCite{Milne2016;textual}{har18}.
#' @param rho (Numeric scalar)
#' Roll-off parameter for harmonic expansion.
#' Defaults to 0.75, after
#' \insertCite{Milne2016;textual}{har18}.
#' @param sigma (Numeric scalar)
#' Standard deviation of the Gaussian smoothing distribution (cents).
#' Defaults to 6.83, after
#' \insertCite{Milne2016;textual}{har18}.
#' @param array_dim (Integerish scalar)
#' Dimensionality of the pitch-class spectrum array.
#' Defaults to 1200, after
#' \insertCite{Milne2016;textual}{har18}.
#' @param ... Arguments passed to specific methods.
#' @return Pitch-class harmonicity, as a numeric scalar.
#' @note This algorithm makes use of \code{\link[hrep]{milne_pc_spectrum}()}
#' as defined in the \code{hrep} package.
#' @md
#' @references
#' \insertAllCited{}
#' @examples
#' pc_harmonicity(c(0, 4, 7))
#' pc_harmonicity(c(0, 3, 7))
#' pc_harmonicity(c(0, 3, 6))
#' @rdname pc_harmonicity
#' @export
pc_harmonicity <- function(x,
                           method = "kl",
                           num_harmonics = 12,
                           rho = 0.75,
                           sigma = 6.83,
                           ...) {
  UseMethod("pc_harmonicity")
}

#' @rdname pc_harmonicity
#' @export
pc_harmonicity.default <- function(x,
                                   method = "kl",
                                   num_harmonics = 12,
                                   rho = 0.75,
                                   sigma = 6.83,
                                   array_dim = 1200,
                                   ...) {
  x <- hrep::pc_set(x)
  do.call(pc_harmonicity, as.list(environment()))
}

#' @rdname pc_harmonicity
#' @export
pc_harmonicity.pc_set <- function(x,
                                  method = "kl",
                                  num_harmonics = 12,
                                  rho = 0.75,
                                  sigma = 6.83,
                                  array_dim = 1200,
                                  ...) {
  checkmate::qassert(method, "S1")

  x <- hrep::milne_pc_spectrum(x,
                               num_harmonics = num_harmonics,
                               rho = rho,
                               sigma = sigma,
                               array_dim = array_dim)
  pc_harmonicity(x,
                 method = method,
                 num_harmonics = num_harmonics,
                 rho = rho,
                 sigma = sigma)
}

#' @rdname pc_harmonicity
#' @export
pc_harmonicity.milne_pc_spectrum <- function(x,
                                             method = "kl",
                                             num_harmonics = 12,
                                             rho = 0.75,
                                             sigma = 6.83,
                                             ...) {
  checkmate::qassert(method, "S1")

  y <- sweep_harmonic_template(x,
                               num_harmonics = num_harmonics,
                               rho = rho,
                               sigma = sigma)
  if (method == "kl") {
    kl_div_from_uniform(y)
  } else if (method == "peak") {
    peak(y)
  } else stop("unrecognised method")
}
